// resin-pine/build/pine.js
// Generated by CoffeeScript 1.12.7

/*
Copyright 2016 Resin.io

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */

/**
 * @module pine
 */
var PinejsClientCore, Promise, assign, defaults, errors, getPine, isEmpty, url, utils,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

assign = require('lodash/assign');

defaults = require('lodash/defaults');

isEmpty = require('lodash/isEmpty');

utils = {
  isString: require('lodash/isString'),
  isNumber: require('lodash/isNumber'),
  isBoolean: require('lodash/isBoolean'),
  isObject: require('lodash/isObject'),
  isArray: require('lodash/isArray'),
  isDate: require('lodash/isDate')
};

url = require('url');

Promise = require('bluebird');

PinejsClientCore = require('pinejs-client/core')(utils, Promise);

errors = require('resin-errors');

getPine = function(arg) {
  var ResinPine, apiKey, apiPrefix, apiUrl, apiVersion, auth, pineInstance, ref, request;
  ref = arg != null ? arg : {}, apiUrl = ref.apiUrl, apiVersion = ref.apiVersion, apiKey = ref.apiKey, request = ref.request, auth = ref.auth;
  apiPrefix = url.resolve(apiUrl, "/" + apiVersion + "/");

  /**
  	 * @class
  	 * @classdesc A PineJS Client subclass to communicate with Resin.io.
  	 * @private
  	 *
  	 * @description
  	 * This subclass makes use of the [resin-request](https://github.com/resin-io-modules/resin-request) project.
   */
  ResinPine = (function(superClass) {
    extend(ResinPine, superClass);

    function ResinPine() {
      return ResinPine.__super__.constructor.apply(this, arguments);
    }


    /**
    		 * @summary Perform a network request to Resin.io.
    		 * @method
    		 * @private
    		 *
    		 * @param {Object} options - request options
    		 * @returns {Promise<*>} response body
    		 *
    		 * @todo Implement caching support.
     */

    ResinPine.prototype._request = function(options) {
      defaults(options, {
        apiKey: apiKey,
        baseUrl: apiUrl
      });
      return auth.hasKey().then(function(hasKey) {
        if (!hasKey && isEmpty(apiKey)) {
          throw new errors.ResinNotLoggedIn();
        }
        return request.send(options).get('body');
      });
    };

    return ResinPine;

  })(PinejsClientCore);
  pineInstance = new ResinPine({
    apiPrefix: apiPrefix
  });
  return assign(pineInstance, {
    API_URL: apiUrl,
    API_VERSION: apiVersion,
    API_PREFIX: apiPrefix
  });
};

module.exports = getPine;
