// resin-device-logs/build/utils.js
// Generated by CoffeeScript 1.12.7

/*
Copyright 2016 Resin.io

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
var assign, getBaseChannel, isArray, isString;

isString = require('lodash/isString');

isArray = require('lodash/isArray');

assign = require('lodash/assign');

getBaseChannel = function(device) {
  return device.logs_channel || device.uuid;
};


/**
 * @summary Get logs channel name for the given device
 * @function
 * @protected
 *
 * @param {Object} device - device
 * @returns {String} logs channel name
 *
 * @example
 * channel = utils.getChannel('...')
 */

exports.getChannel = function(device, suffix) {
  if (suffix == null) {
    suffix = 'logs';
  }
  return "device-" + (getBaseChannel(device)) + "-" + suffix;
};


/**
 * @summary Get logs and clear logs channel names for the given device
 * @function
 * @protected
 *
 * @param {Object} device - device
 * @returns {Object} { channel, clearChannel }
 *
 * @example
 * channel = utils.getChannel('...')
 */

exports.getChannels = function(device) {
  return {
    channel: exports.getChannel(device),
    clearChannel: exports.getChannel(device, 'clear-logs')
  };
};


/**
 * @summary Extract messages from PubNub payload
 * @function
 * @public
 *
 * @param {*} message - message
 * @returns {Object[]} log messages
 *
 * @example
 * messages = utils.extractMessages('foo bar')
 */

exports.extractMessages = function(message) {
  if (isString(message)) {
    return [
      {
        isSystem: /\[system\]/.test(message),
        message: message,
        timestamp: null,
        serviceId: null
      }
    ];
  } else if (isArray(message)) {
    return message.map(function(arg) {
      var c, m, s, t;
      m = arg.m, t = arg.t, s = arg.s, c = arg.c;
      return {
        message: m,
        timestamp: t,
        isSystem: Boolean(s),
        serviceId: c != null ? c : null
      };
    });
  } else {
    return [
      assign({
        isSystem: false,
        timestamp: null,
        serviceId: null
      }, message)
    ];
  }
};
